File: frontend/src/types/solver.ts
// solver.ts
export interface SolverResult {
  solve_id: string;
  status: 'OPTIMAL' | 'FEASIBLE' | 'INFEASIBLE';
  objective_value: number;
  optimality_gap: number;
  wall_time_sec: number;
  train_schedules: Record<string, { start: number; end: number; delay: number }>;
  block_schedules: Record<
    string,
    {
      start: number;
      end: number;
      section: string;
      is_shadow: boolean;
      shadow_parent?: string;
    }
  >;
  clashes_detected: number;
  shadow_merges: number;
  xai: XAIData;
}

export interface XAIData {
  conflict_resolutions: { block_id: string; shifted_minutes: number; reason: string }[];
  shadow_detections: { primary: string; shadow: string; time_saved_hours: number }[];
  constraint_waterfall: { train_delay_pct: number; block_deviation_pct: number; shadow_bonus_pct: number };
}

export interface GrantedBlock {
  id: number;
  solve_id: string;
  demand_id: number;
  demand_code: string;
  granted_start_minutes: number;
  granted_end_minutes: number;
  section_from: string;
  section_to: string;
  is_shadow: boolean;
  shadow_parent_id?: number;
}

File: backend/app/schemas/solver.py
from pydantic import BaseModel
from typing import List, Dict, Any, Optional

class TrainScheduleOutput(BaseModel):
    start: int
    delay: int

class ConflictResolution(BaseModel):
    block_id: str
    shifted_minutes: int
    reason: str

class ShadowDetection(BaseModel):
    primary: str
    shadow: str
    time_saved_hours: float

class XAIResponse(BaseModel):
    conflict_resolutions: List[ConflictResolution]
    shadow_detections: List[ShadowDetection]

class TelemetryPoint(BaseModel):
    iteration: int
    objective_cost: float
    best_bound: float
    time_sec: float

class SolveRequest(BaseModel):
    demands: Optional[List[Dict[str, Any]]] = None
    chaos_mode: bool = False
    section: str = "BINA-ITARSI"
    max_solve_time_sec: float = 8.0

class SolveResponse(BaseModel):
    solve_id: str
    status: str
    optimality_gap: float
    wall_time_sec: float
    clashes_detected: int
    shadow_merges: int
    train_schedules: Dict[str, TrainScheduleOutput]
    xai: XAIResponse
    telemetry: List[TelemetryPoint]
